README.md
# Guexit.ImageGeneration

[![codecov](https://codecov.io/gh/Guexit/Guexit.ImageGeneration/branch/main/graph/badge.svg?token=U47OQU1RG5)](https://codecov.io/gh/Guexit/Guexit.ImageGeneration)

AI Image Generation Service

[View Changelog](CHANGELOG.md)

## Local

### Setup

1. Install [Python](https://www.python.org/downloads/) (look for 3.10 version) if you haven't already. You can also install [Pyenv](https://github.com/pyenv/pyenv) to manage your Python versions.

2. Install Poetry from [here](https://python-poetry.org/docs/#installation).

3. If you're on Linux or Windows, install CUDA 11.7.

    - First, install your NVIDIA Drivers.
    - Then, install CUDA:

        - If in WSL, execute: `sudo apt install nvidia-cuda-toolkit`

4. Initialize a new Poetry environment and install dependencies:

    ```shell
    poetry shell
    poetry install
    ```

    If you want to develop locally, you can install poetry with the following command:

    ```shell
    poetry install --with dev
    ```

5. Install CUDA if you're in Linux or Windows:

    Ideally, we want to install CUDA 11.7.

    - First of all, install your NVIDIA Drivers.
    - Then, install CUDA:

        - If in WSL, execute: `sudo apt install nvidia-cuda-toolkit`

6. Install pre-commit:

    ```shell
    pre-commit install
    ```

### Run

1. In one terminal, start the server with:

    ```shell
    sh start_server.sh
    ```

2. You can go to [http://127.0.0.1:5000/docs](http://127.0.0.1:5000/docs) to see the swagger API. You will find an example of a valid request. It should look like this:

    ```json
    {
        "text_to_image": {
            "model_path": "digiplay/Juggernaut_final",
            "model_scheduler": "dpmsolver_multistep",
            "prompt": {
                "positive": "portrait of samantha prince set in fire, cinematic lighting, photorealistic, ornate, intricate, realistic, detailed, volumetric light and shadow, hyper HD, octane render, unreal engine insanely detailed and intricate, hypermaximalist, elegant, ornate, hyper-realistic, super detailed",
                "negative": "bad quality, malformed",
                "guidance_scale": 16.5
            },
            "height": 688,
            "width": 512,
            "num_inference_steps": 50,
            "num_images": 2,
            "seed": 57857
        }
    }
    ```

3. You can also execute the script in the example folder to test the API:

```shell
python3 examples/txt2image.py --model_path "digiplay/Juggernaut_final" /
                               --model_scheduler dpmsolver_multistep /
                               --positive_prompt "portrait of samantha prince set in fire, cinematic lighting, photorealistic, ornate, intricate, realistic, detailed, volumetric light and shadow, hyper HD, octane render, unreal engine insanely detailed and intricate, hypermaximalist, elegant, ornate, hyper-realistic, super detailed"
                               --negative_prompt "bad quality, malformed"
                               --guidance_scale 16.5
                               --height 688
                               --width 512
                               --num_inference_steps 50
                               --num_images 2
                               --seed 57857
```

### Using ImageGenerationMessageHandler

The ImageGenerationMessageHandler in `scripts/image_generation_message_handler.py` is a script that processes incoming messages to generate images, uploads them to Azure Blob Storage, and sends a message with the generated image URLs to an Azure Service Bus topic.

To use the ImageGenerationMessageHandler, you need to have a few environment variables set up:

```shell
AZURE_STORAGE_CONNECTION_STRING: The connection string for your Azure Storage account.
AZURE_SERVICE_BUS_CONNECTION_STRING: The connection string for your Azure Service Bus instance.
AZURE_STORAGE_CONTAINER_NAME: The name of the container in your Azure Storage account where the images will be uploaded.
AZURE_SERVICE_BUS_QUEUE_NAME: The name of the queue in your Azure Service Bus instance where messages will be consumed.
AZURE_SERVICE_BUS_TOPIC_NAME: The name of the topic in your Azure Service Bus instance where messages will be published.
```

These are in `env_samples.env`, but you need to have them in a `.env` file in the project directory: `cp env_samples.env .env` and then substitute the values with the correct ones.

You can set these environment variables in your terminal or in a .env file in the project directory. Here's an example of how to set them in your terminal:

```shell
export AZURE_STORAGE_CONNECTION_STRING="<your_storage_connection_string>"
export AZURE_SERVICE_BUS_CONNECTION_STRING="<your_service_bus_connection_string>"
export AZURE_STORAGE_CONTAINER_NAME="<your_storage_container_name>"
export AZURE_SERVICE_BUS_QUEUE_NAME="<your_service_bus_queue_name>"
export AZURE_SERVICE_BUS_TOPIC_NAME="<your_service_bus_topic_name>"
```

Once you have the environment variables set up, you first need to have the service running on one terminal:

```shell
sh start_server.sh
```

Then you can start consuming messages from Azure Service Bus with:

```shell
sh start_generating.sh
```

This will start the ImageGenerationMessageHandler and begin processing messages from the specified Azure Service Bus queue. The generated images will be uploaded to the specified Azure Storage container, and the resulting image URLs will be sent as a message to the specified Azure Service Bus topic.

## Docker

To run the service and the handler in Docker:

1. Have Docker installed.

2. Git credentials must be stored in `~/.git-credentials`. To do so please execute:

    ```shell
    git config --global credential.helper "store --file ~/.git-credentials"
    ```

3. Make sure you've copied the `env_samples.env` into `.env` with the following command:

    ```shell
    cp env_samples.env .env
    ```

    MAKE SURE TO CHANGE THE ENV VARIABLES ACCORDINGLY.

4. Use Docker Compose to build the image and run the container:

    ```shell
    docker compose up --build -d
    ```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.6.2] - 2023-12-25

### Added

- Adds tags to service bus message

## [0.6.1] - 2023-12-25

### Added

- GitHub action to make a release, with a changelog and a tag.

## [0.6.0] - 2023-12-24

### Added

- `cloud_manager` module while removing the existing repository.

## [0.5.0] - 2023-12-07

### Added

- [stabilityai/sdxl-turbo](https://huggingface.co/stabilityai/sdxl-turbo) model which is x10 faster than [segmind/SSD-1B](https://huggingface.co/segmind/SSD-1B).

### Changed

- Pytorch version from `==2.0.0` to `>=2.1.1`.

## [0.4.0] - 2023-12-05

### Added

- Probability to variables in styles.

## [0.3.0] - 2023-11-11

### Added

- `test_prompt_randomness` feature for improved prompt generation.
- Plot variable distributions utils for quickly displaying prompt variable value distribution across generated prompts.

### Changed

- Public domain characters now used for famous characters variables.
- Cleaned up codebase, removing duplicates and unused styles.

### Fixed

- Updated Docker GitHub action and codecov settings.

## [0.2.0] - 2023-10-24

### Added

- Safety checker with retry mechanism for NSFW content detection.
- Tests for the new safety checker.

### Fixed

- Fixed seed on general style.

## [0.1.3] - 2023-10-23

### Added

- Around x1.5 speedup on inference.
- Famous characters and new elements to styles.

### Changed

- Switched to Python 3.10.


examples/txt2image.py
import io
import zipfile

from image_generation.utils import call_image_generation_api


def extract_zip_images(response):
    # Extract zip images
    z = zipfile.ZipFile(io.BytesIO(response.content))
    z.extractall("images/")


if __name__ == "__main__":
    import argparse

    # Text to image input args
    parser = argparse.ArgumentParser()
    parser.add_argument("--model_path", type=str, default="stabilityai/sdxl-turbo")
    parser.add_argument("--model_scheduler", type=str, default="euler_a")
    parser.add_argument(
        "--positive_prompt",
        type=str,
        default="3d 32-bit isometric anime ice cream stand with an old woman",
    )
    parser.add_argument(
        "--negative_prompt",
        type=str,
        default="",
    )
    parser.add_argument("--guidance_scale", type=float, default=0.0)
    parser.add_argument("--height", type=int, default=688)
    parser.add_argument("--width", type=int, default=512)
    parser.add_argument("--num_inference_steps", type=int, default=2)
    parser.add_argument("--num_images", type=int, default=1)
    parser.add_argument("--seed", type=int, default=-1)
    args = parser.parse_args()
    # Call Image Generation API
    response = call_image_generation_api(
        host="http://127.0.0.1:5000",
        endpoint="/text_to_image",
        request_object={
            "model_path": args.model_path,
            "model_scheduler": args.model_scheduler,
            "prompt": {
                "positive": args.positive_prompt,
                "negative": args.negative_prompt,
                "guidance_scale": args.guidance_scale,
            },
            "height": args.height,
            "width": args.width,
            "num_inference_steps": args.num_inference_steps,
            "num_images": args.num_images,
            "seed": args.seed,
        },
    )
    # Save images
    extract_zip_images(response)
